test/metabase/lib/test_util/mocks_31769.cljc

```diff
@@ -0,0 +1,89 @@
+                        (lib/with-join-conditions [(lib/suggested-join-condition q people)]))))
+      (lib/breakout q (let [breakout (m/find-first #(and (= (:id %) (id-fn :products :category))
+                                                         (not= (:lib/source %) :source/implicitly-joinable))
+                                                   (lib/breakoutable-columns q))]
+                        (assert breakout)
+                        breakout))
+      (lib/aggregate q (lib/count)))))
+
+(defn card-2-query
+  "For reproducing #31769: create a query against `products` with a breakout on `products.category` and a `count`
+  aggregation."
+  [metadata-provider id-fn]
+  (let [products (lib.metadata/table metadata-provider (id-fn :products))]
+    (as-> (lib/query metadata-provider products) q
+      (lib/breakout q (let [breakout (m/find-first #(and (= (:id %) (id-fn :products :category))
+                                                         (not= (:lib/source %) :source/implicitly-joinable))
+                                                   (lib/breakoutable-columns q))]
+                        (assert breakout)
+                        breakout)))))
+
+(defn mock-metadata-provider
+  "For reproducing #31769: Create a composed metadata provider with two Cards based on [[card-1-query]]
+  and [[card-2-query]]. This is mostly for MLv2 usage rather than QP usage, at least until the
+  QP-use-MLv2-metadata-providers PR lands. Until that lands, QP tests will need to use `with-temp` to create the
+  Cards."
+  ([]
+   (mock-metadata-provider meta/metadata-provider meta/id 1 2))
+
+  ([metadata-provider id card-1-id card-2-id]
+   (lib/composed-metadata-provider
+    (lib.tu/mock-metadata-provider
+     {:cards [{:id            card-1-id
+               :name          "Card 1"
+               :database-id   (id)
+               :dataset-query (card-1-query metadata-provider id)}
+              {:id            card-2-id
+               :name          "Card 2"
+               :database-id   (id)
+               :dataset-query (card-2-query metadata-provider id)}]})
+    metadata-provider)))
+
+(defn query
+  "For reproducing #31769: create a query using a `:source-card` with [[card-1-query]] as its source, joining a Card
+  with [[card-2-query]]."
+  ([]
+   (query (mock-metadata-provider meta/metadata-provider meta/id 1 2)
+          1
+          2))
+
+  ([metadata-provider card-1-id card-2-id]
+   (let [card-1 (lib.metadata/card metadata-provider card-1-id)
+         card-2 (lib.metadata/card metadata-provider card-2-id)]
+     (assert card-1)
+     (assert card-2)
+     (as-> (lib/query metadata-provider card-1) q
+       (lib/append-stage q)
+       (lib/join q (-> (lib/join-clause card-2)
+                       (lib/with-join-conditions [(let [lhs (m/find-first #(= (:name %) "CATEGORY")
+                                                                          (lib/join-condition-lhs-columns q card-2 nil nil))
```

Comment:
```suggestion
                                                                          (lib/join-condition-lhs-columns q card-1 nil nil))
```

Reply:
The args for `join-condition-lhs-columns` are 

```clj
(join-condition-lhs-columns query joinable lhs-column-or-nil rhs-column-or-nil)
```

and `card-2` is the thing we're joining.